web_programming/get_google_search_results.py

```diff
@@ -0,0 +1,125 @@
+    True
+    >>> len(parse_results("")) == 0
+    True
+    """
+    params = {"q": str(query)}
+    response = None
+    try:
+        response = requests.get(BASE_URL + "/search", headers=HEADERS, params=params)
+    except requests.exceptions.RequestException:
+        return []
+    css_identifier_result = ".tF2Cxc"
+    css_identifier_title = "h3"
+    css_identifier_link = ".yuRUbf a"
+    css_identifier_text = "#rso .lyLwlc"
+
+    soup = BeautifulSoup(response.text, "lxml")
+    results = soup.select(css_identifier_result)
+    output = []
+    for result in results:
+        title = result.select_one(css_identifier_title).text
+        link = result.select_one(css_identifier_link)["href"]
+        try:
+            snippet = result.select_one(css_identifier_text).text
+        except Exception:
+            snippet = None
+        item = {"title": title, "link": link, "text": snippet}
+        output.append(item)
+    stat_table = BeautifulSoup(response.content, "html.parser").find("table")
+    stat_table = soup.find("table")
+    if stat_table is None:
+        return []
+    table_data = stat_table.select("td a")
+    next_page = []
+
+    for item in table_data:
+        next_page.append(item["href"])
+    for next_page_link in next_page:
+        new_link = BASE_URL + next_page_link
+        try:
+            response = requests.get(new_link, headers=HEADERS)
+
+        except requests.exceptions.RequestException:
+            continue
+        soup = BeautifulSoup(response.text, "lxml")
+        results = soup.select(css_identifier_result)
+        for result in results:
+            title = result.select_one(css_identifier_title).text
+            link = result.select_one(css_identifier_link)["href"]
+            try:
+                snippet = result.select_one(css_identifier_text).text
+            except Exception:
+                snippet = None
+            item = {"title": title, "link": link, "text": snippet}
+            output.append(item)
+    return output
+
+
+def write_google_search_results(query: str = "", filename: str = "") -> str:
+    """Writes the Google search result page for the given query into a local file.
+    Args:
+        query: The search term provided by the user. Default is "potato".
+        filename: The name of the file into which the search results should be
+            saved.
+
+    Returns:
+        str: The name of the file into which the search results were saved.
+
+    >>> write_google_search_results("python", "test") != None
+    True
+    >>> write_google_search_results("", "tet.html") != None
+    True
+    >>> write_google_search_results("python", "") != None
+    True
+    >>> write_google_search_results("", "") != None
+    True
+    >>> "test" in write_google_search_results("python", "test")
+    True
+    >>> "test1" in write_google_search_results("", "test1")
+    True
+    >>> "potato" in write_google_search_results("", "")
+    True
```

Comment:
Make the doctests more strict by testing the outputs for equality:
```suggestion
    >>> write_google_search_results("python", "test") == "test.txt"
    True
    >>> write_google_search_results("", "tet.html") == "tet.html.txt"
    True
    >>> write_google_search_results("python", "") == "python.txt"
    True
    >>> write_google_search_results("", "test1") == "test1.txt"
    True
    >>> write_google_search_results("", "") == "potato.txt"
    True
```